### DISCUSSION TITLE
Documentation of llama 2 model architecture details?

### DISCUSSION THREAD

**Discussion by dougdew64** [ggml-org/llama.cpp#3788] [General]
Where is the llama 2 model architecture documentation which was used to figure out what code to write for the llama.cpp file?

This llama doc does not seem to provide the kind of details which were apparently important when writing the code in the llama.cpp file: https://arxiv.org/pdf/2302.13971.pdf

My goal is to gain a thorough understanding of the code in the llama.cpp file.

Thank you in advance.

**Comment by dougdew64** [comment]
This llama doc also does not seem to provide all of the architectural details necessary to write the code in the llama.cpp file: https://arxiv.org/pdf/2307.09288.pdf

**Reply by slaren** [reply]
For any missing details, you can use the python implementations as a reference, such as https://github.com/facebookresearch/llama/ or HF Transformers.

**Reply by dougdew64** [reply]
Thank you very much @slaren

**Reply by dougdew64** [reply]
@slaren I'm attempting to implement a SwiftUI app which provides a visualization of the model graph. My very, very limited understanding of the code is that the structure of a llama 2 model graph is known in advance and is baked into the llama.cpp code. In other words, the graph structure is not discovered when loading a *.gguf file. Is my understanding correct? If so, then it seems that I should be able to implement a model visualizer which makes assumptions about the graph structure.

**Reply by slaren** [reply]
That's mostly correct, the graphs are for the most part predefined in the code. This is the function that creates the graph for llama models:
https://github.com/ggerganov/llama.cpp/blob/6961c4bd0b5176e10ab03b35394f1e9eab761792/llama.cpp#L3091
llama.cpp supports other models as well, and each of them have a similar `llm_build` function to create the computation graph.

Note that ggml already has an utility to export a visualization of a graph in graphviz dot file format. You can create it by uncommenting these lines, but you may also want to restrict the graph building code to just 1 layer, otherwise the graph will be huge.
https://github.com/ggerganov/llama.cpp/blob/6961c4bd0b5176e10ab03b35394f1e9eab761792/llama.cpp#L6027-L6030

**Reply by dougdew64** [reply]
Also, please pardon any of my misuses of terminology. I am new to this, and so am still very ignorant. For example, while reading a comment in the ggml.h file just now, I noticed the phrase "computation graph." So, perhaps my earlier use of "model graph" should have instead been "computation graph."

**Reply by dougdew64** [reply]
Thank you VERY much for pointing out that vizualization export function. You have been super helpful @slaren!

**Reply by dougdew64** [reply]
I used that graphviz export function to generate a llama.dot file. Then, I started the dot command to generate a llama.png file from the llama.dot file. So far, the dot command has been running for two hours at 100% cpu. Your advice to limit to just one layer is looking better by the hour. ;-)

![image](https://github.com/ggerganov/llama.cpp/assets/98898390/9609092c-f0bb-421a-9b2d-8b31bc3b9431)

**Reply by dougdew64** [reply]
Thirty hours of cpu time, and still no result from graphviz dot. It's probably fair to say that attempting to use dot to generate a *.png file from the export of a complete graph is a poor idea. 
![image](https://github.com/ggerganov/llama.cpp/assets/98898390/bb983663-8952-472e-870e-68770966d76f)
